rdrf/rdrf/admin.py

```diff
@@ -221,14 +139,175 @@ def has_delete_permission(self, request, *args, **kwargs):
+                return None
+            else:
+                logger.info("Exported YAML Data for %s OK" % registry.name)
+            return yaml_data
+        except Exception as ex:
+            logger.error("export registry action for %s error: %s" % (registry.name, ex))
+            messages.error(request, "Custom Action Failed: %s" % ex)
+            return None
+
+    def export_registry_action(self, request, registry_models_selected):
+        from datetime import datetime
+        export_time = str(datetime.now())
+
+        registries = list(registry_models_selected)
+
+        if len(registries) == 1:
+            registry = registries[0]
+            yaml_export_filename = registry.name + ".yaml"
+            yaml_data = self.export_registry(registry, request)
+            if yaml_data is None:
+                return
+
+            yaml_file = io.StringIO()
+            yaml_file.write(yaml_data)
+            yaml_file.flush()
+            yaml_file.seek(0)
+
+            response = HttpResponse(FileWrapper(yaml_file), content_type='text/yaml')
+            yaml_export_filename = "export_%s_%s" % (export_time, yaml_export_filename)
+            response['Content-Disposition'] = 'attachment; filename="%s"' % yaml_export_filename
+
+            return response
+        else:
+            import zipfile
+            zipped_file = io.BytesIO()
+            zf = zipfile.ZipFile(zipped_file, mode='w', compression=zipfile.ZIP_DEFLATED)
+
+            for registry in registries:
+                yaml_data = self.export_registry(registry, request)
+                if yaml_data is None:
+                    return
+
+                zf.writestr(registry.code + '.yaml', yaml_data)
+
+            zf.close()
+            zipped_file.flush()
+            zipped_file.seek(0)
+
+            response = HttpResponse(FileWrapper(zipped_file), content_type='application/zip')
+            name = "export_" + export_time + "_" + \
+                   reduce(lambda x, y: x + '_and_' + y, [r.code for r in registries]) + ".zip"
+            response['Content-Disposition'] = 'attachment; filename="%s"' % name
+
+            return response
+    export_registry_action.short_description = _("Export")
+
+    def generate_questionnaire_action(self, request, registry_models_selected):
+        for registry in registry_models_selected:
+            registry.generate_questionnaire()
+    generate_questionnaire_action.short_description = _("Generate Questionnaire")
+
+    @staticmethod
+    def missing_cache_table():
+        for cache_alias in settings.CACHES:
+            cache = caches[cache_alias]
+            if hasattr(cache, '_table') and cache._table not in connection.introspection.table_names():
+                return cache._table
+
+    def setup_registration_view(self, request, registry_code):
+        registry = Registry.objects.get(code=registry_code)
+
+        missing_table = self.missing_cache_table()
```

Comment:
Could be using the walrus operator here, but we're still waiting for flake8 to support it

Reply:
I **love** the walrus!